[AggressiveInstcombine] Fold away shift in or reduction chain.

davemgreen opened the pull request:
If we have `icmp eq or(a, shl(b)), 0` then the shift can be removed so long as it is nuw or nsw. It is still comparing that some bits are non-zero. 
https://alive2.llvm.org/ce/z/nhrBVX.

This is also true of ne, and true for longer or chains.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7742,6 +7765,11 @@ Instruction *InstCombinerImpl::visitICmpInst(ICmpInst &I) {
     }
   }
 
+  // icmp eq/ne or(shl(a), b), 0 -> icmp eq/ne or(a, b)
```
```suggestion
  // icmp eq/ne or(shl(a), b), 0 -> icmp eq/ne or(a, b), 0
```

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7742,6 +7765,11 @@ Instruction *InstCombinerImpl::visitICmpInst(ICmpInst &I) {
   }
 
+  // icmp eq/ne or(shl(a), b), 0 -> icmp eq/ne or(a, b)
+  if (I.isEquality() && match(Op1, m_Zero()) && isa<Instruction>(Op0))
```
It would be better to move this fold into `InstCombinerImpl::foldICmpWithZero`.

davemgreen replied:
Thanks - I wasn't sure which of foldICmpEquality, foldICmpBinOp, foldICmpWithZero, foldICmpWithConstant or.. foldICmpInstWithConstant (?) it would live best in. I've moved it to the end of foldICmpWithZero (and had to move the function earlier in the file).

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -5046,6 +5046,29 @@ static Instruction *foldICmpOrXX(ICmpInst &I, const SimplifyQuery &Q,
 
+static Value *foldShiftAwayFromOrChain(Instruction &I,
+                                       InstCombiner::BuilderTy &Builder) {
+  if (I.getOpcode() != Instruction::Or)
```
Missing one-use check.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -5046,6 +5046,29 @@ static Instruction *foldICmpOrXX(ICmpInst &I, const SimplifyQuery &Q,
+    return Builder.CreateOr(A, B);
+
+  Value *Op0 = I.getOperand(0);
+  if (isa<Instruction>(Op0))
```
We can use `dyn_cast` instead of `isa + cast`.

dtcxzyw reviewed (commented):
I am fine with this fold itself. But it prevents AggressiveInstCombine from merging consecutive loads: https://godbolt.org/z/aq94MeM3n

We should not perform this fold when the shift's LHS is `zext(load)`.
If you think this workaround is too fragile, an alternative solution is to convert `(load iN p) | (load iN p+N/8) == 0` into `(load i2N p) == 0` in AggressiveInstCombine.

davemgreen commented:
I've tried to move this into aggressive instcombine to allow it to check for trees of loads first. The transform is looking up through trees already, so maybe that is a better place for it. Let me know what you think.

dtcxzyw reviewed (approved):
LGTM. Thank you!

davemgreen committed:
[AggressiveInstcombine] Fold away shift in or reduction chain. #137875

If we have icmp eq or(a, shl(b)), 0 then the shift can be removed so long as it
is nuw or nsw. It is still comparing that some bits are non-zero.
https://alive2.llvm.org/ce/z/nhrBVX.

This is also true of ne, and true for longer or chains.
